tower/src/filter/predicate.rs

```diff
@@ -1,25 +1,69 @@
-use super::error::Error;
 use std::future::Future;
 
-/// Checks a request
-pub trait Predicate<Request> {
+/// Checks a request asynchronously.
+pub trait AsyncPredicate<Request> {
     /// The future returned by `check`.
-    type Future: Future<Output = Result<(), Error>>;
+    type Future: Future<Output = Result<Self::Request, Self::Error>>;
+
+    /// The type of requests returned by `check`.
+    ///
+    /// This request is forwarded to the inner service if the predicate
+    /// succeeds.
+    type Request;
+    /// The type of errors returned by `check`.
+    ///
+    /// If the predicate fails, this is returned rather than calling the inner
+    /// service.
+    type Error: Into<crate::BoxError>;
```

Comment:
What's the value of having an `Error` type parameter when they're always coerced to the box error? What's the downside to having the traits simply require that boxed errors are returned?

Reply:
hmm, yeah, we could just remove that, good point! the only downside is that the user has to box the error, which adds a bit of code to the predicate implementation. with this approach, we can automatically box the error.

on the other hand, if the user code returns errors via `?`, they don't have to write `Box::new` etc either...

Reply:
Right. And the `.map_err(Into::into)` may be less onerous than having extra `type Error = MyError` defns?

I don't have a strong opinion on this either way, but I wanted to raise the question to get some consensus on which pattern is better.